src/terminal.rs

```diff
@@ -0,0 +1,240 @@
+        writeln!(
+            self.stdout,
+            "\nSua mensagem: {}",
+            style(&todo.message).green()
+        )
+        .map(|_| ())
+        .map_err(TerminalError::Stdout)
+    }
+
+    fn show_todos(&mut self, prefix: &str) -> Result<(), TerminalError> {
+        let current_elements = self.todos.list.len();
+        let mut message = String::from(prefix);
+
+        if current_elements == 0 {
+            let empty_prefix = "Lista de TODOs está vazia.".to_owned();
+            message.clear();
+            message.push_str(&empty_prefix);
+        }
+
+        for (index, todo) in self.todos.list.iter().enumerate() {
+            message.push_str(&format!("{} - {}, ", index, todo.message));
+        }
+
+        writeln!(self.stdout, "{}", style(message).yellow().bold(),)
+            .map_err(TerminalError::Stdout)?;
+        Ok(())
+    }
+
+    fn read_input(&mut self) -> Result<String, TerminalError> {
+        let mut buf = String::new();
+
+        self.stdin
+            .read_line(&mut buf)
+            .map_err(TerminalError::Stdin)?;
+
+        Ok(buf.trim().to_string())
+    }
+
+    fn show_menu(&mut self) -> Result<(), TerminalError> {
+        writeln!(
+            self.stdout,
+            "{}",
+            style("\n1 - Adicionar novo Todo \n2 - Remover Todo \n3 - Listar Todos \n4 - Esvaziar lista \n5 - Sair\n").green()
+        )
+        .map(|_| ())
+        .map_err(TerminalError::Stdout)
+    }
+
+    fn option_map(&mut self) -> HashMap<usize, UserOption> {
+        let mut option_map = HashMap::new();
+
+        option_map.insert(1, UserOption::NewTodo);
+        option_map.insert(2, UserOption::RemoveTodo);
+        option_map.insert(3, UserOption::ShowList);
+        option_map.insert(4, UserOption::RemoveAllTodos);
+        option_map.insert(5, UserOption::Quit);
+
+        return option_map;
+    }
+
+    fn select_index(&mut self) -> Result<UserOption, TerminalError> {
+        let input = self.read_input()?;
+
+        match input.parse::<usize>() {
+            Ok(parsed_input) => {
+                if parsed_input > self.todos.list.capacity() + 1 || parsed_input < 1 {
+                    return Err(TerminalError::InvalidOption);
+                }
+                match self.option_map().get(&parsed_input) {
+                    Some(user_option) => Ok(user_option.clone()),
+                    None => Err(TerminalError::InvalidOption),
+                }
+            }
+            Err(_) => Err(TerminalError::InvalidOption),
+        }
+    }
+
+    fn select_from_list(&mut self) -> Result<usize, TerminalError> {
+        let input = self.read_input()?;
+
+        match input.parse::<usize>() {
+            Ok(parsed_input) => {
+                for (i, _) in self.todos.list.iter().enumerate() {
+                    if i == parsed_input {
+                        return Ok(i);
+                    }
+                }
```

Comment:
Ao invés desse `for`, você pode apenas verificar se o índice que o usuário passou excede o comprimento da lista de TODOs atual. Você pode utilizar o `len()` para isso.

A vantagem é que não precisaremos de percorrer a lista para verificar se o número é válido ou não. Em termos de "complexidade assintótica", a comparação é constante. :)